orttraining/orttraining/training_ops/cuda/tensor/gather_elements_grad.cc

```diff
@@ -0,0 +1,135 @@
+          rank,
+          buffer_input_dims,
+          buffer_input_strides,
+          indices_data,
+          indices_size,
+          buffer_indices_dims,
+          fdm_indices_strides,
+          reinterpret_cast<const CudaT*>(update_data),
+          axis,
+          reinterpret_cast<CudaT*>(output_data));
+    } else if (utils::IsPrimitiveDataType<int64_t>(Tin_type)) {
+      const int64_t* indices_data = indices_tensor->template Data<int64_t>();
+      return GatherElementsGradImpl(
+          rank,
+          buffer_input_dims,
+          buffer_input_strides,
+          indices_data,
+          indices_size,
+          buffer_indices_dims,
+          fdm_indices_strides,
+          reinterpret_cast<const CudaT*>(update_data),
+          axis,
+          reinterpret_cast<CudaT*>(output_data));
+    }
+
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED, "Type for Tin is not supported yet in ScatterElements.");
+  }
+};
+
+Status GatherElementsGrad::ComputeInternal(OpKernelContext* context) const {
+  const Tensor* shape = context->Input<Tensor>(0);
+  const TensorShape data_shape(shape->template Data<int64_t>(), shape->Shape().Size());
+
+  const int axis = static_cast<int>(HandleNegativeAxis(axis_, data_shape.NumDimensions()));
+
+  const auto* indices_tensor = context->Input<Tensor>(1);
+  const auto* updates_tensor = context->Input<Tensor>(2);
+
+  const auto& indices_dims = indices_tensor->Shape().GetDims();
+  const int64_t indices_size = indices_tensor->Shape().Size();
+  const auto& updates_dims = updates_tensor->Shape().GetDims();
+  if (indices_dims.size() != updates_dims.size()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Indices and updates must have the same rank");
+  }
+
+  for (size_t i = 0; i < indices_dims.size(); ++i) {
+    if (indices_dims[i] != updates_dims[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices vs updates dimensions differs at position=", i,
+                             " ", indices_dims[i], " vs ", updates_dims[i]);
+    }
+  }
+
+  // According to the spec the rank of ind/upd shall be the same as input(data)
+  // and we also want to make sure that the dimensions of the of the ind/upd do not
+  // exceed that of the input
+  const auto& input_dims = data_shape.GetDims();
+  if (input_dims.size() != indices_dims.size()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices must have the same rank as Input. Indices rank=",
+                           indices_dims.size(), ". Input rank=", input_dims.size());
+  }
+
+  for (size_t i = 0; i < input_dims.size(); ++i) {
+    if (input_dims[i] < indices_dims[i]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices dim=", indices_dims[i], " at pos=", i,
+                             " is greater than input dim=", input_dims[i]);
+    }
+  }
+
+  int rank = (int)input_dims.size();
+  Tensor* output_tensor = context->Output(0, data_shape);
```

Comment:
>output_tensor [](start = 10, length = 13)

Need to memset the output buffer to zero first, as not every element in the output will be touched

Reply:
thanks for catching this!!!!